packages/smart-wallet/src/offerWatcher.js

```diff
@@ -0,0 +1,267 @@
+
+          state.walletHelper.updateStatus(state.status, state.address);
+        },
+        onNewContinuingOffer(
+          offerId,
+          invitationAmount,
+          invitationMakers,
+          publicSubscribers,
+        ) {
+          const { state } = this;
+
+          void state.walletHelper.addContinuingOffer(
+            offerId,
+            invitationAmount,
+            invitationMakers,
+            publicSubscribers,
+            state.address,
+          );
+        },
+
+        publishResult(result) {
+          const { state, facets } = this;
+
+          const passStyle = passStyleOf(result);
+          // someday can we get TS to type narrow based on the passStyleOf result match?
+          switch (passStyle) {
+            case 'bigint':
+            case 'boolean':
+            case 'null':
+            case 'number':
+            case 'string':
+            case 'symbol':
+            case 'undefined':
+              facets.helper.updateStatus({ result });
+              break;
+            case 'copyRecord':
+              if ('invitationMakers' in result) {
+                // save for continuing invitation offer
+
+                void facets.helper.onNewContinuingOffer(
+                  String(state.status.id),
+                  state.invitationAmount,
+                  result.invitationMakers,
+                  result.publicSubscribers,
+                );
+              }
+              facets.helper.updateStatus({ result: UNPUBLISHED_RESULT });
+              break;
+            default:
+              // drop the result
+              facets.helper.updateStatus({ result: UNPUBLISHED_RESULT });
+          }
+        },
+
+        exitOpenSeat() {
+          const { state } = this;
+          void E.when(E(state.seatRef).hasExited(), hasExited => {
+            if (!hasExited) {
+              void E(state.seatRef).tryExit();
+            }
+          });
+        },
+      },
+
+      /** @type {OutcomeWatchers['paymentWatcher']} */
+      paymentWatcher: {
+        async onFulfilled(payouts) {
+          const { state, facets } = this;
+
+          // This will block until all payouts succeed, but user will be updated
+          // since each payout will trigger its corresponding purse notifier.
+          const amountPKeywordRecord = objectMap(payouts, paymentRef =>
+            E.when(paymentRef, payment => state.deposit.receive(payment)),
+          );
+          const amounts = await deeplyFulfilledObject(amountPKeywordRecord);
+          facets.helper.updateStatus({ payouts: amounts });
+        },
+        /**
+         * @param {Error} err
+         * @param {UserSeat} seat
+         */
+        onRejected(err, seat) {
+          const { facets } = this;
+
+          if (isUpgradeDisconnection(err)) {
+            void watchForPayout(facets, seat);
+          } else {
+            facets.helper.updateStatus({ error: err.toString() });
+            facets.helper.exitOpenSeat();
+          }
```

Comment:
these three rejection handlers have the same logic save for the callback. please DRY.

e.g.
```suggestion
          facets.helper.handleError(err, seat, watchForPayout);
```

Reply:
also, isn't the `updateStatus` redundant with the one in `smartWallet.js`? Please confirm whether these are necessary here.

I'm pretty sure both the logging and the seat exiting are handled in smartWallet's error handler. If so, then this is just,
```suggestion
          if (isUpgradeDisconnection(err)) {
            void watchForPayout(facets, seat);
          }
```
In which case it's better not to make a helper. It's pretty DRY as is.

Reply:
correct on all counts.